openedx_blockstore_relay/transfer_data.py

```diff
@@ -0,0 +1,257 @@
+
+        if scrape_assets:
+            self._add_static_assets(data)
+
+        return response
+
+    def _post_bundle_file(self, data, path, name=None, content_type='application/octet-stream', public=True):
+        """
+        POST the given file to the Blockstore bundle.
+
+        Arguments:
+        * data: string/byte data, or an open file pointer (required)
+        * path: path to the file (required)
+        * name: base file name, appended to path if provided (optional)
+        * content_type: file content type (optional)
+        * public: default True
+        """
+        if name is None:
+            name = os.path.basename(path)
+        else:
+            path = os.path.join(path, name)
+        files = [('data', (name, data, content_type))]
+        metadata = {
+            'path': path,
+            'public': public,
+        }
+        LOG.debug("POST file %s to Bundle <%s>...", metadata, self.bundle_uuid)
+        response = requests.post(self.bundle_files_url, files=files, data=metadata)
+        response.raise_for_status()
+        return response
+
+    def _add_static_assets(self, data):
+        """
+        Store the static assets paths, and fully-qualified URLs from the given data string in our static assets list.
+        """
+        static_paths = self._scrape_static_paths(data)
+        self.assets.extend(static_paths)
+
+    def _scrape_static_paths(self, data):
+        """
+        Pull the static assets (paths, and fully-qualified URLs) from the given data string.
+        """
+        from static_replace import replace_static_urls  # pylint: disable=import-error
+        static_paths = []
+        replace_static_urls(data,
+                            course_id=self.block_key.course_key,
+                            static_paths=static_paths)
+        return static_paths
+
+
+class TransferUnit(TransferBlock):
+    """
+    Transfers an Unit (vertical) and any child XBlocks from Open edX to Blockstore.
+    """
+
+    def upload_block(self):
+        """
+        Upload the block and its child components to the bundle.
+        """
+        LOG.debug('Transfer %s and children to Bundle <%s>', self.block_key, self.bundle_uuid)
+        block = self.get_block()
+        olx_node = ElementTree.Element(self.block_key.block_type)
+        for child in block.get_children():
+            LOG.debug('Transfer child %s to Bundle <%s>', child.location, self.bundle_uuid)
+            filesystem = MemoryFS()
+            try:
+                child_olx = child.definition_to_xml(filesystem)
+            except NotImplementedError:
+                LOG.warning("Child block type not supported, skipping (%s)", child.location)
+                continue
+
+            self.upload_components(filesystem)
+            # FIXME -- either upload the child OLX, or append it to the root node.  Not both.
+            # self.upload_olx(child_olx, child)
+            olx_node.append(child_olx)
```

Comment:
All the olx files should be separate in Blockstore.

Reply:
Thought of another reason why we have to maintain separate OLX files and just reference the children instead of inlining them -- Directed Acyclic Graphs!

Reply:
@pomegranited 

> Thought of another reason why we have to maintain separate OLX files and just reference the children instead of inlining them -- Directed Acyclic Graphs!

Wait, which way do you mean that? For courses, we would like to enforce trees only and prohibit DAGs (by which I mean we don't want to allow DAGs where one block has multiple parents). They caused too many problems. 

https://openedx.atlassian.net/wiki/spaces/EDUCATOR/pages/34734086/Why+is+the+Course+Structure+a+DAG

https://openedx.atlassian.net/browse/TNL-4922

That means that if a course is a DAG, trying to import it into Blockstore as-is should fail with an error.

*Or* we could allow DAGs but if a block appears in multiple places, it gets treated by the LMS as two separate blocks. I think that's actually a better approach and supports adaptive learning and a bunch of other features. In that case, the fix would be in the XBlock runtime when it generates the full usage_key for each block, ensuring that the usage_key is based in part on the location of the block in the course.

Reply:
@bradenmacdonald 

Lol.. now I have no idea what this function should do when it encounters a DAG 😄 

Currently, the content that gets imported into the Bundle is as close to the Open edX Studio export as I could manage without creating a `runtime`.  The import loop checks to make sure that it's not recursing infinitely through DAGs, but it allows them without error, just like Studio does.  

We'd still need these checks in place to throw an error when a DAG is encountered, so if we decide later to disallow DAGs, we can do that.  Or if we decide that the DAGs should be allowed, but we want the sub-blocks inlined in a single OLX file that uses XIncludes or some other way to reference internal blocks, we can do that too.

Then how the LMS chooses to read all of this -- that's also up for discussion 😄

Reply:
haha ok, thanks @pomegranited